reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -319,492 +244,554 @@ public String toString() {
+		final Clock                               clock;
+		final Deque<BoundedState>                 idleQueue;
+		final PriorityBlockingQueue<BoundedState> busyQueue;
+
+		//constructor for SHUTDOWN
+		private BoundedServices() {
+			this.parent = null;
+			this.clock = Clock.fixed(Instant.EPOCH, ZoneId.systemDefault());
+			this.busyQueue = new PriorityBlockingQueue<>();
+			this.idleQueue = new ConcurrentLinkedDeque<>();
+		}
+
+		BoundedServices(BoundedElasticScheduler parent) {
 			this.parent = parent;
-			if (parent != null) {
-				this.exec = Schedulers.decorateExecutorService(parent, parent.get());
-			}
-			else {
-				this.exec = Executors.newSingleThreadScheduledExecutor();
-				this.exec.shutdownNow();
+			this.clock = parent.clock;
+			this.busyQueue = new PriorityBlockingQueue<>(parent.maxThreads,
+					Comparator.comparingInt(bs -> bs.markCount));
+			this.idleQueue = new ConcurrentLinkedDeque<>();
+		}
+
+		/**
+		 * Trigger the eviction by computing the oldest acceptable timestamp and letting each {@link BoundedState}
+		 * check (and potentially shutdown) itself.
+		 */
+		void eviction() {
+			final long evictionTimestamp = parent.clock.millis();
+			List<BoundedState> idleCandidates = new ArrayList<>(idleQueue);
+			for (BoundedState candidate : idleCandidates) {
+				if (candidate.tryEvict(evictionTimestamp, parent.ttlMillis)) {
+					idleQueue.remove(candidate);
+					decrementAndGet();
+				}
 			}
 		}
 
-		@Override
-		public void dispose() {
-			if (exec != null) {
-				if (this != SHUTDOWN && !parent.shutdown) {
-					//in case of work, re-create an ActiveWorker
-					DeferredFacade deferredFacade = parent.deferredFacades.poll();
-					if (deferredFacade != null) {
-						deferredFacade.setService(this);
+		/**
+		 * Pick a {@link BoundedState}, prioritizing idle ones then spinning up a new one if enough capacity.
+		 * Otherwise, picks an active one by taking from a {@link PriorityQueue}. The picking is
+		 * optimistically re-attempted if the picked slot cannot be marked as picked.
+		 *
+		 * @return the picked {@link BoundedState}
+		 */
+		BoundedState pick() {
+			for (;;) {
+				int a = get();
+				if (a == DISPOSED) {
+					return CREATING; //synonym for shutdown, since the underlying executor is shut down
+				}
+
+				if (!idleQueue.isEmpty()) {
+					//try to find an idle resource
+					BoundedState bs = idleQueue.pollLast();
+					if (bs != null && bs.markPicked()) {
+						busyQueue.add(bs);
+						return bs;
 					}
-					else {
-						//if no more work, the service is put back at end of the cached queue and new expiry is started
-						CachedServiceExpiry e = new CachedServiceExpiry(this,
-								System.currentTimeMillis() + parent.ttlSeconds * 1000L);
-						parent.idleServicesWithExpiry.offerLast(e);
-						if (parent.shutdown) {
-							if (parent.idleServicesWithExpiry.remove(e)) {
-								exec.shutdownNow();
-							}
-						}
+					//else optimistically retry
```

Comment:
WDYT about using `continue` here? Otherwise, if we accidentally add some logic to the end of this loop, it will get executed when inner `if` resolves to false

Reply:
avoiding that is the goal of the comment, but I'll augment the comment with `"(implicit continue here)"`. otherwise triggers inspection "continue is unnecessary as the last statement in a loop".